text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+            //      reference, since the reference borrows `self`.
+            unsafe { pointer.as_ref() }
+        })
+    }
+
+    pub fn into_left(mut self) -> Option<core::boxed::Box<L, S>> {
+        self.is_left.then(|| {
+            let handle = self.handle;
+
+            //  Safety:
+            //  -   `self.store` will no longer be used.
+            let store = unsafe { ManuallyDrop::take(&mut self.store) };
+
+            mem::forget(self);
+
+            //  Safety:
+            //  -   `handle` was allocated by `store`.
+            //  -   `handle` is valid.
+            //  -   The block of memory associated to `handle` contains a live instance of `L`.
+            unsafe { core::boxed::Box::from_raw_parts(handle, store) }
+        })
+    }
+
+    //  And implementations of `as_left_mut`, `right`, `as_right`, `as_right_mut`, ...
+}
+
+impl<L, R, S: Store> Drop for Either<L, R, S> {
+    fn drop(&mut self) {
+        //  Safety:
+        //  -   `handle` was allocated by `store`.
+        //  -   `handle` is still valid.
+        let pointer = unsafe { self.store.resolve(self.handle) };
+
+        if self.is_left {
+            let pointer: *mut L = pointer.cast().as_ptr();
+
+            //  Safety:
+            //  -   `pointer` is valid for both reads and writes.
+            //  -   `pointer` is properly aligned.
+            unsafe { ptr::drop_in_place(pointer) }
+        } else {
+            let pointer: *mut R = pointer.cast().as_ptr();
+
+            //  Safety:
+            //  -   `pointer` is valid for both reads and writes.
+            //  -   `pointer` is properly aligned.
+            unsafe { ptr::drop_in_place(pointer) }
+        };
+
+        let layout = if self.is_left {
+            Layout::new::<L>()
+        } else {
+            Layout::new::<R>()
+        };
+
+        //  Safety:
+        //  -   `self.store` will no longer be used.
+        let store = unsafe { ManuallyDrop::take(&mut self.store) };
+
+        //  Safety:
+        //  -   `self.handle` was allocated by `self.store`.
+        //  -   `self.handle` is still valid.
+        //  -   `layout` fits the block of memory associated to `self.handle`.
+        unsafe { store.deallocate(self.handle, layout) }
+    }
+}
+```
+
+By using `Store`, I empower my users to use my type in a wide variety of scenarios, some of which I cannot even fathom.
+
+The overhead of using `Store` over `Allocator` is also fairly low, so that the added flexibility comes at little to no
+cost to me.
+
+More examples of collections can be found at https://github.com/matthieu-m/storage/tree/main/src/collection, including
+a complete linked list, a box draft, a concurrent vector draft, and a skip list draft.
+
+
+##  Store Implementer Guide
+
+The API of `Store` requires internal mutability, and that's it. I can otherwise provide as few or as many guarantees as
+I wish.
```

Comment:
Do we, actually, lose `!noalias`, or was it a chimera to start with?

### In-line stores

I do not think that any in-line store can reasonably provide `!noalias` guarantees. I must admit having a more than shaky understanding from `!noalias` (I'm barely LLVM literate) so I am going from the assumption that it mirrors C's `__restrict`.

As a generalizable example, let's consider `[MaybeUninit<T>; 2]`. For such a store, `&mut self` is a trap.

The problem lies in the fact that having handed out a first slice of memory covering the first element of the array, that slice may currently be resolved to a `&T` or `&mut T`, and while this reference is actively used, taking a `&mut self` to the entire array is instant UB.

This means that in the case of in-line stores, `&mut self` is restricted to single allocation stores, or stores which invalidate any resolved pointers on any method call (even `resolve`, if the objective is to allow LRU stores without internal mutability).

### Off-line stores

The issue about in-line stores is actually more general that it seems.

An off-line store using `Box` or `Vec`, for example, to store its state, would be remiss to mutably access the content of such collection via a `&mut self` to the collection as that implies a `__restrict` to the entire content, running headlong into the same UB as the in-line store should any resolved reference exist.

Stores that do not use `&mut self` (such a `Global`) do not have the issue... but do not benefit from `&mut` either, so that's hardly motivating.

Once again, only a single allocation store would realistically benefit... but single allocation stores by nature have very little state to care about in the first place.

### Footgun

As far as I am concerned, this makes any API using `&mut self` a serious footgun risk. An implementer may be tempted to just take advantage of this to "bypass" the need for interior mutability, and run headlong into UB.

That is, for any multiple allocations store, the actual memory to partition into allocated memory blocks _must_ be accessed solely by pointers to avoid implying shared/exclusive access as `&T`/`&mut T` tend to, and in the case of multiple allocations in-line stores (or others using `Box`/`Vec` which propagate uniqueness), _must_ be kept within an `UnsafeCell` as otherwise taking a reference to the store itself triggers UB.

That's a serious disadvantage, which would require thorough documentation. And even then...

### Does the mutability of the Store matter?

As far as I understand, `__restrict` keyword can be applied to any resolved `&mut T` just fine, and thus the only mutable state that matters here is the internal meta-data state of the `Store`.

I am not sure if the absence of `__restrict` access to this internal meta-data state matters that much, optimization-wise.

Reply:
The main issue is that `noalias`/`__restrict` is at the scope of the entire pointer. If you have `b: &Box<u8, InlineStore<u8>>` and do `**b; ffi(); **b;`, it doesn't matter that references to just the box content are known `noalias`, because the parent pointer isn't, and the known-`noalias` reference is recreated from the potentially aliasing parent pointer on either side of the FFI.

And yes, part of taking `&mut self` *is* that doing so invalidates any extant resolved pointers. It's not an "issue" if the store then takes that `&mut` and projects it to outline storage, because the entire purpose of providing `&mut` was to allow the storage to do so.

Yes, using `MultipleStore` correctly if it takes `&mut self` is difficult; it's the exact same difficulty as mutably accessing multiple indices of a `Vec` at the same time, or that pushing or popping from a vector (at least theoretically[^1]) invalidates any pointers to vector items, even if no reallocation occurs, despite many developers not liking that behavior.

[^1]: The current `Vec` implementation doesn't actually manifest a full `&mut [T]` during push/pop, so doesn't actually cause pointer invalidation, but it would be permitted to. Additionally, at the LLVM level, only the statically sized prefix is actually subject to restrictions if not actually accessed, and the statically sized prefix of a slice is 0 bytes (for an empty slice).

Reply:
Mitigating the potential footgun was the exact reason that v2 returns lifetime-having values; any access to or reference of the storage which would invalidate the lifetime invalidates the resolved pointer. The lifetimes weren't needless restrictions; they were very deliberate communication tools for actual invalidation timing.

Reply:
> This means that in the case of in-line stores, &mut self is restricted to single allocation stores,

To be frank, I still don't quite see the use case for inline multiple storage. The storage is necessarily a single block since, so providing it as a single chunk/slice seems desirable. Most cases want `Vec<T>`, and if you actually want a linked list, it's probably for the ability to split/join/splice lists, so you want `SharingStore`, and/or you want to insert/remove/split/splice from just a single pointer to a link in the list (and not also the base), so you want each link to own a (ideally zero-sized) allocator handle.

The remaining potential use case is mixed size/type allocation. I don't see the concrete use case where doing so is desirable, because one object necessarily owns all of these separate little allocations, in order to be able to clean them up on drop, but if one object is managing all of them, why not manage them all as a single allocation block?

When multiple allocations are involved, that generally means multiple allocator handles. Those handles being either references of some storage provider or a ZST directing you to a `static` storage provider.

Reply:
> The remaining potential use case is mixed size/type allocation. I don't see the concrete use case where doing so is desirable, because one object necessarily owns all of these separate little allocations, in order to be able to clean them up on drop, but if one object is managing all of them, why not manage them all as a single allocation block?

I can definitely see a case of using `BTreeSet`, `BTreeMap`, `HashSet`, and/or `HashMap` with in-line memory store.

I am not sure whether `HashSet`/`HashMap` today uses a single or multiple allocations, but I am certain that `BTreeSet` and `BTreeMap` do.